Hooks, improvements, and hacks to puter.com, and implementation of a puter integration

dsagal opened the pull request:
grist-static changes:
- Pass Python errors back through pyCall
- Support importing Excel files by supporting binary data and choosing the right parser.
- Various hooks to allow overriding or extending grist-static behavior.
- Hack to add a "save" hook for when "Save document" button is clicked.

puter.com integration, via page/index_puter.html:
- Support opening and saving .grist files.
- Support opening all file types supported for imports (csv, xlsx, etc)
- Support saving csv files back to the original location, with a warning.
- Show logged-in Puter user (not important, but we don't have UI to avoid showing any user)
- Show when the file is saved/unsaved, and prompt to save when needed.

dsagal committed:
Implement opening blank doc or from file; showing Puter user; and catching click to save

dsagal committed:
Implement saving to puter (though not yet opening), and silence pointless 'Script error'

dsagal committed:
Implement opening saved .grist files

dsagal committed:
Move out puter-specific JS code to puter.js; implement renames

dsagal committed:
Implement reasonable behavior for opening and saving Grist docs

dsagal committed:
Support importing Excel files; choose the right parser, and support binary data. Also a workaround for puter read() bug

dsagal committed:
Pass through Python errors back through pyCall

dsagal committed:
When importing, consider the doc unchanged (saved) until the first change

dsagal committed:
Support saving open CSVs back to the original CSV file

dsagal committed:
Silence spurious uncaught errors without needing a change to grist-core. Also, don't consider an unchanged empty file unsaved

dsagal committed:
Remove incorrect comments

dsagal commented:
There are no tests... I manually tested some flows in Puter, but there is a high risk this breaks something with other uses of grist-static, like the csv viewer. Might be safer to release to a separate URL for the sake of Puter, and leave a proper update of grist-static for when someone has time to poke at grist-static a bit more?

dsagal committed:
Include page/*.js files into build-dist, to get puter.js

dsagal commented:
I released for now into `s3://grist-static/experiments/ePUTER01/`, with the entrypoint at `index_puter2.html` (the "2" suffix because i couldn't get past caching of a bad version). That's enough to get the Puter grist app working, so I'll actually update it to that (vs the current pointer to Grist SaaS on prod), but happy to update if/when it's available in a more permanent place.

dsagal committed:
Hide/disable disfunctional elements with CSS; add a script to publish a puter version

dsagal committed:
Fix save behavior for new files

paulfitz reviewed (approved):
Very neat! Impressed you were able to work on this shaky foundation and make it do what you want. Also very happy to see grist-static put to good use! Thanks a lot Dmitry!

I skimmed over all the code and liked what I see. Would be strange to do a rigorous review given the underlying shenanigans :)

dsagal commented:
Thanks :) Two notes:
1. I removed `ext/app/server/lib/SQLite.ts` entirely because `ext/app/server/lib/SqliteJs.ts` seems the only one used, and it was confusing to have both. I assumed one copy was accidental or obsolete. But let me know if I misunderstood why there are two.
2. I added `scripts/push_puter_version.sh` script. It seems easier to keep "puter" releases separate from "grist-static" releases even if they reuse code, so there is less chance of puter vs csv-viewer updates breaking each other.

paulfitz commented:
> I assumed one copy was accidental or obsolete. But let me know if I misunderstood why there are two.

My memory is too vague on this. There was an intermediate step of bootstrapping the project, before there was an actual fully working thing, where maybe sqlite.js was in use but from node?? No I don't remember. Fine to delete if everything works.
puter script is good idea :+1:
